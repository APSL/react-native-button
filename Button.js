'use strict';

var React = require('react-native');
var {
  View,
  TouchableOpacity,
  Text,
  StyleSheet,
  PropTypes,
  ActivityIndicatorIOS,
  ProgressBarAndroid,
  Platform
} = React;

var Button = React.createClass({
  propTypes: Object.assign({},
    TouchableOpacity.propTypes,
    {textStyle: PropTypes.object,
    children: PropTypes.string.isRequired,
    isLoading: PropTypes.bool,
    isDisabled: PropTypes.bool,
    activityIndicatorColor: PropTypes.string},
  ),

  _renderInnerText: function () {
    if (this.props.isLoading) {
      if (Platform.OS !== 'android') {
        return (
          <ActivityIndicatorIOS
            animating={true}
            size='small'
            style={styles.spinner}
            color={this.props.activityIndicatorColor || 'black'}
          />
        );
      } else {
        return (
          <ProgressBarAndroid
            style={[{
              height: 20,
            }, styles.spinner]}
            styleAttr='Inverse'
          />
        );
      }
    }
    return (
      <Text style={[styles.textButton, this.props.textStyle]}>
        {this.props.children}
      </Text>
    );
  },

  render: function () {
    // Extract TouchableOpacity props
    var touchableProps = {
      onPress: this.props.onPress,
      onPressIn: this.props.onPressIn,
      onPressOut: this.props.onPressOut,
      onLongPress: this.props.onLongPress
    };

    if (this.props.isDisabled === true || this.props.isLoading === true) {
      return (
        <View style={[styles.button, this.props.style, styles.opacity]}>
          {this._renderInnerText()}
        </View>
      );
    } else {
      return (
        <TouchableOpacity {...touchableProps}
          style={[styles.button, this.props.style]}>
          {this._renderInnerText()}
        </TouchableOpacity>
      );
    }
  }
});

var styles = StyleSheet.create({
  button: {
    height: 44,
    flexDirection: 'row',
    borderWidth: 1,
    borderRadius: 8,
    marginBottom: 10,
    alignSelf: 'stretch',
    justifyContent: 'center',
  },
  textButton: {
    fontSize: 18,
    alignSelf: 'center',
  },
  spinner: {
    alignSelf: 'center',
  },
  opacity: {
    opacity: 0.5,
  },
});

module.exports = Button;
