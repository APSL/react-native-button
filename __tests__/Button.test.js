/* global test */

import { Text, View } from 'react-native'
import React from 'react'
import Button from '../Button'
import renderer from 'react-test-renderer'

describe('Button', () => {
  test('Renders', () => {
    const component = renderer.create(
      <Button>
        Button
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Renders disabled', () => {
    const component = renderer.create(
      <Button isDisabled={true}>
        Disabled button
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Renders loading', () => {
    const component = renderer.create(
      <Button isLoading={true}>
        Loading button
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Renders custom activity indicator', () => {
    const component = renderer.create(
      <Button isLoading={true} activityIndicator={<Text>Loading</Text>}>
        Custom loader button
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Renders with a inner View', () => {
    const component = renderer.create(
      <Button>
        <View />
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('Should contain children', () => {
    const component = renderer.create(
      <Button>
        Press me!
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    const props = component.getInstance().props
    expect(props.children).toEqual('Press me!')
  })
  test('Should react to the onPress event', () => {
    const mockFunction = jest.fn(() => 'Mock function')
    const component = renderer.create(
      <Button onPress={mockFunction}>
        Press me!
      </Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    const props = component.getInstance().props
    expect(props.onPress()).toEqual('Mock function')
    expect(mockFunction).toBeCalled()
  })
})
