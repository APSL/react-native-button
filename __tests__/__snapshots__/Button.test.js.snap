// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 44,
      "justifyContent": "center",
      "marginBottom": 10,
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "fontSize": 18,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`Button Renders disabled 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 44,
        "justifyContent": "center",
        "marginBottom": 10,
      },
      undefined,
      Object {
        "opacity": 0.5,
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "fontSize": 18,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Disabled button
  </Text>
</View>
`;

exports[`Button Renders loading 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 44,
        "justifyContent": "center",
        "marginBottom": 10,
      },
      undefined,
      Object {
        "opacity": 0.5,
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="black"
    hidesWhenStopped={true}
    size="small"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  />
</View>
`;

exports[`Button Renders with a inner View 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 44,
      "justifyContent": "center",
      "marginBottom": 10,
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View />
</View>
`;

exports[`Button Should contain children 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 44,
      "justifyContent": "center",
      "marginBottom": 10,
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "fontSize": 18,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Press me!
  </Text>
</View>
`;

exports[`Button Should react to the onPress event 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 44,
      "justifyContent": "center",
      "marginBottom": 10,
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "fontSize": 18,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Press me!
  </Text>
</View>
`;
